L46-core-tls-credential-API.md

```diff
@@ -0,0 +1,463 @@
+                                 unsigned int refresh_interval_sec);
+
+  // Constructor to get credential updates from root file path only.
+  FileWatcherCertificateProvider(const std::string& root_cert_path,
+                                 unsigned int refresh_interval_sec);
+    }
+
+
+```
+### Custom Verification
+We aim to provide distinct interfaces for chain building customization and for additional validation of an already built chain. These two features have distinctly different expected users. Doing some extra validation based on the contents of the peer certificate chain is an operation that anyone doing (m)TLS might be interested in, and it should not and does not require any deep X.509 expertise. Replacing how chain building works is a relatively rare requirement and it does require deep X.509 expertise to do correctly. We want extra validation to be easy and clear for users to add without worrying about interacting with chain building itself. On the other hand, chain building is complex to implement, and we suspect that most users will stay with the default behavior. However, to enable advanced use, such as dynamic selection of the trust bundle based on the peer certificate (e.g. SPIFFE federation), this is needed. 
+#### Post Handshake Verification
+```c++
+// Contains the verification-related information associated with a connection
+// request. Users should not directly create or destroy this request object, but
+// shall interact with it through CertificateVerifier's Verify() and Cancel().
+// TODO(gtcooke94) Add expected formats for all string values here.
+class TlsCustomVerificationCheckRequest {
+ public:
+  explicit TlsCustomVerificationCheckRequest(
+      grpc_tls_custom_verification_check_request* request);
+  ~TlsCustomVerificationCheckRequest() {}
+
+  absl::string_view target_name() const;
+  absl::string_view peer_cert() const;
+  absl::string_view peer_cert_full_chain() const;
+  absl::string_view common_name() const;
+  // The subject name of the root certificate used to verify the peer chain
+  // If verification fails or the peer certificate is self-signed, this will be an
+  // empty string. If verification is successful, it is a comma-separated list,
+  // where the entries are of the form "FIELD_ABBREVIATION=string"
+  // ex: "CN=testca,O=Internet Widgits Pty Ltd,ST=Some-State,C=AU"
+  // ex: "CN=GTS Root R1,O=Google Trust Services LLC,C=US"
+  absl::string_view verified_root_cert_subject() const;
+  std::vector<absl::string_view> uri_names() const;
+  std::vector<absl::string_view> dns_names() const;
+  std::vector<absl::string_view> email_names() const;
+  std::vector<absl::string_view> ip_names() const;
+
+}
+
+// The base class of all internal verifier implementations, and the ultimate
+// class that all external verifiers will eventually be transformed into.
+// To implement a custom verifier, do not extend this class; instead,
+// implement a subclass of ExternalCertificateVerifier. Note that custom
```

Comment:
Why do we need the extra "ExternalCertificateVerifier" layer? Can we explain that here? Should we add documentation for the "ExternalCertificateVerifier" class itself, similar to the others?

Reply:
Oh good catch, I removed all of the code for the external layer but missed this comment

The external layer is not needed anymore.